packages/@aws-cdk/aws-codebuild/test/test.codebuild.ts

```diff
@@ -1153,6 +1232,28 @@ export = {
       }
     });
 
+    test.done();
+  },
+
+  'filter groups validation'(test: Test) {
+    const stack = new cdk.Stack();
+
+    // should throw exception when webhook is not specified but filterGroups is
+    test.throws(() => {
+      new codebuild.Project(stack, 'Project', {
+        source: new codebuild.GitHubSource({
+          owner: 'testowner',
+          repo: 'testrepo',
+          cloneDepth: 3,
+          filterGroups: [
+            [
+              { type: WebhookFilterType.Event, pattern: 'PUSH' }
+            ]
+          ],
+        })
+      });
+    }, Error, "filterGroups property could only be set when webhook property is true");
```

Comment:
I would actually change this behavior. I think if `filterGroups` was specified, and `webhook` was not, I think `webhook` should default to `true`. Only specifying `filterGroups` **and** `webhook: false` should result in an exception.

What do you think?

Reply:
I'm thinking it in a different way. If customers want to remove filter groups but keep webhook enabled, they don't need to add `webhook: true` which is a cognitive load for customers.

Reply:
Right. So when a dilemma like this happens, I like to compare the number of customers in each case. Here, you have every single person who wants to use filter groups, vs. a group that first uses filter groups, and then stops using them. It's pretty clear that the first group is larger (as the second group is a subset of the first one), so I think it makes more sense to optimize the experience for the larger group.

Reply:
I think we should follow the principle of least astonishment here. Customers won't be surprised to see an error if they specify filterGroups but don't specify webhook. However, they will feel surprised and confused when removing filterGroups causes their webhook being deleted. Could you please give me an example of somewhere else doing this kind of override behavior to help me better understand this case? I couldn't think of any at this moment.

Reply:
> Could you please give me an example of somewhere else doing this kind of override behavior to help me better understand this case? I couldn't think of any at this moment.

Sure. As an example, in CodeDeploy's `ServerDeploymentGroup`, we change the value of the rollback alarm configuration depending on whether the customer has provided any alarms (see [here](https://github.com/awslabs/aws-cdk/blob/ed39a8c4c00d5261bf6deb17134ed3f81d75cf63/packages/%40aws-cdk/aws-codedeploy/lib/rollback-config.ts#L23) and [here](https://github.com/awslabs/aws-cdk/blob/ed39a8c4c00d5261bf6deb17134ed3f81d75cf63/packages/%40aws-cdk/aws-codedeploy/lib/server/deployment-group.ts#L214)).